### PR TITLE
feat/cody: Brings image modality for BYOK users

### PR DESCRIPTION
**Pull request by PriNova**
This PR brings image modality for BYOK users via the Google LLM provider.

The PR is behind the `cody.dev.models` experimental feature flag. You need to configure it in the settings.json like this:

```
"cody.dev.models": [
    {
            "provider": "google",
            "model": "gemini-2.0-flash-exp",
            "inputTokens": 1048576,
            "outputTokens": 8192,
            "apiKey": "your_key_goes_here",
            "options": {
                "temperature": 0.0
            }
        },
 ]
 ```

https://github.com/user-attachments/assets/96b15374-84b4-427e-9156-6d683cca7a7a

Model Selection overhauled:

Screenshot 2024-12-16 105822

## Test plan

Build Cody based on this PR

### Manual Testing Steps
1. Model Selection:
   - Open Cody chat
   - Click on model selector
   - Verify Gemini Flash 2.0 model shows as Vision model

2. Image Upload Flow:
   - Select a Gemini Flash 2.0 model
   - Verify image selection button is visible in toolbar
   - Click image button
   - Select an image file

3. Chat Interaction:
   - With uploaded image, send a message
   - Check response includes image context

4. Edge Cases:
   - Switch between models and verify image selection button visibility
   - Test with various image formats ( jpeg, png, webg )

5. Drag 'n' Drop:
   - Switch to Gemini Flash 2.0
   - Drag an image file from a folder into the Chat Box (need to press SHIFT while dragging https://github.com/microsoft/vscode/issues/182449)
   - Check that image selection button changes to filename

## Notes
- Feature only available for Gemini Flash 2.0 model

### Added

- Add Gemini Flash 2.0 experimental vision model support via cody dev models flag [pull/6354](https://github.com/sourcegraph/cody/pull/6354)

### CODE REVIEW

File: vscode/src/chat/chat-view/ChatBuilder.ts [hunk 1]
```diff
@@ -94,7 +95,8 @@ export class ChatBuilder {
 
         public readonly sessionID: string = new Date(Date.now()).toUTCString(),
         private messages: ChatMessage[] = [],
-        private customChatTitle?: string
+        private customChatTitle?: string,
+        private images: ImageData[] = []
```
[the remark this answers is not in the document]
Comment (PriNova) [reply vscode/src/chat/chat-view/ChatBuilder.ts:99]: I have also thought about this and find this a great idea. Not only would the user have visual feedback, but there would also be multiple media blobs available in the future.
In its current state, however, this would collide with the other non-visual models, which would mean adding an additional context filter later on (totally feasible).

Additionally, I'm just not sure if a large number of images in the chat history would hurt performance. Depending on the specification of the computer, a slowdown in chat history management is observed: https://linear.app/sourcegraph/issue/CODY-4516/vscode-cody-extension-lags-with-large-chat-history-40-items

**Comment by ykdojo** [comment]
Tried it again, and it looks great! Copying and pasting is still not working for me, though
